clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -267,12 +267,15 @@ public synchronized void awaitUpdate(final int lastVersion, final long maxWaitMs
      * @param topics
      */
     public synchronized void setTopics(Collection<String> topics) {
-        Set<TopicPartition> toRemove = lastSeenLeaderEpochs.keySet()
+        Set<TopicPartition> partitionsToRemove = lastSeenLeaderEpochs.keySet()
                 .stream()
                 .filter(tp -> !topics.contains(tp.topic()))
                 .collect(Collectors.toSet());
-        toRemove.forEach(lastSeenLeaderEpochs::remove);
-        toRemove.forEach(cache::removePartitionInfo);
+        partitionsToRemove.forEach(lastSeenLeaderEpochs::remove);
+
+        Set<String> topicsToRemove = partitionsToRemove.stream().map(TopicPartition::topic).collect(Collectors.toSet());
```

Comment:
The set here is built using `lastSeenLeaderEpochs`, which will only be populated for recent versions which return leader epoch information. However, it seems generally useful to keep the cache consistent with the `topics` collection. Possibly one option is to change the API to `MetadataCache.retain(Set<String> topics)`?